tools/criticalstat.py

```diff
@@ -0,0 +1,327 @@
+                    help="Duration in uS (microseconds) below which we filter")
+
+args = parser.parse_args()
+
+preemptoff = False
+irqoff = False
+
+if args.irqoff:
+    preemptoff = False
+    irqoff = True
+elif args.preemptoff:
+    preemptoff = True
+    irqoff = False
+
+debugfs_path = subprocess.Popen ("cat /proc/mounts | grep -w debugfs" +
+    " | awk '{print $2}'",
+    shell=True,
+    stdout=subprocess.PIPE).stdout.read().split("\n")[0]
+
+if debugfs_path == "":
+    print("ERROR: Unable to find debugfs mount point");
+    sys.exit(0);
+
+trace_path = debugfs_path + "/tracing/events/preemptirq/";
+
+if (not os.path.exists(trace_path + "irq_disable") or
+   not os.path.exists(trace_path + "irq_enable") or
+   not os.path.exists(trace_path + "preempt_disable") or
+   not os.path.exists(trace_path + "preempt_enable")):
+    print("ERROR: required tracing events are not available\n" +
+        "Make sure the kernel is built with CONFIG_DEBUG_PREEMPT " +
+        "and CONFIG_PREEMPTIRQ_EVENTS enabled. Also please disable " +
+        "CONFIG_PROVE_LOCKING and CONFIG_LOCKDEP on older kernels.")
+    sys.exit(0)
+
+bpf_text = """
+#include <uapi/linux/ptrace.h>
+#include <linux/sched.h>
+
+enum addr_offs {
+    START_CALLER_OFF,
+    START_PARENT_OFF,
+    END_CALLER_OFF,
+    END_PARENT_OFF
+};
+
+struct start_data {
+    u32 addr_offs[2];
+    u64 ts;
+    int idle_skip;
+    int active;
+};
+
+struct data_t {
+    u64 time;
+    u64 stack_id;
+    u32 cpu;
+    u64 id;
+    u32 addrs[4];   /* indexed by addr_offs */
+    char comm[TASK_COMM_LEN];
+};
+
+BPF_STACK_TRACE(stack_traces, 16384);
+BPF_PERCPU_ARRAY(sts, struct start_data, 1);
+BPF_PERCPU_ARRAY(isidle, u64, 1);
+BPF_PERF_OUTPUT(events);
+
+/*
+ * In the below code we install tracepoint probes on preempt or
+ * IRQ disable/enable critical sections and idle events, the cases
+ * are combinations of 4 different states.
+ * The states are defined as:
+ * CSenter: A critical section has been entered. Either due to
+ *          preempt disable or irq disable.
+ * CSexit: A critical section has been exited. Either due to
+ *         preempt enable or irq enable.
+ * Ienter: The CPU has entered an idle state.
+ * Iexit: The CPU has exited an idle state.
+ *
+ * The scenario we are trying to detect is if there is an overlap
+ * between Critical sections and idle entry/exit. If there are any
+ * such cases, we avoid recording those critical sections since they
+ * are not worth while to record and just add noise.
+ */
+TRACEPOINT_PROBE(power, cpu_idle)
+{
+    int idx = 0;
+    u64 val;
+    struct start_data *stdp, std;
+
+    stdp = sts.lookup(&idx);
+    bpf_probe_read(&std, sizeof(struct start_data), stdp);
+
+    // Mark active sections as that they should be skipped
+
+    // Handle the case CSenter, Ienter, CSexit, Iexit
+    // Handle the case CSenter, Ienter, Iexit, CSexit
+    if (std.active) {
+        std.idle_skip = 1;
+        sts.update(&idx, &std);
+    }
```

Comment:
So, the issue you saw in [#1794](https://github.com/iovisor/bcc/pull/1794/files#r193598376) is due to a known bug/limitation of the verifier. It's documented at #347. The [d71962f](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=d71962f3e627b5941804036755c844fabfb65ff5) kernel patch should fix it for 4.18.

In the meantime, we'll need to copy to the stack. Could you try the following?
```c
stdp = sts.lookup(&idx);
if (stdp && stdp->active) {
    std = *stdp;
    std.idle_skip = 1;
    sts.update(&idx, &std);
}
```
We might want to comment this so that we remember to change it in the future, when the patch becomes widespread enough.

Reply:
Here's the diff addressing both the comments ontop of my previous patch, let me know if it looks Ok to you. I am testing it now. Once testing looks good I'll update PR
```
diff --git a/tools/criticalstat.py b/tools/criticalstat.py
index 2b60411aaf6b..0f90c8438699 100755
--- a/tools/criticalstat.py
+++ b/tools/criticalstat.py
@@ -124,14 +124,19 @@ TRACEPOINT_PROBE(power, cpu_idle)
     u64 val;
     struct start_data *stdp, std;
 
-    stdp = sts.lookup(&idx);
-    bpf_probe_read(&std, sizeof(struct start_data), stdp);
-
     // Mark active sections as that they should be skipped
 
     // Handle the case CSenter, Ienter, CSexit, Iexit
     // Handle the case CSenter, Ienter, Iexit, CSexit
-    if (std.active) {
+    stdp = sts.lookup(&idx);
+    if (stdp && stdp->active) {
+        /*
+         * Due to verifier issues, we have to copy contents
+         * of stdp onto the stack before the update.
+         * Fix it to directly update once kernel patch d71962f
+         * becomes more widespread.
+         */
+        std = *stdp;
         std.idle_skip = 1;
         sts.update(&idx, &std);
     }
@@ -195,7 +200,7 @@ TRACEPOINT_PROBE(preemptirq, TYPE_enable)
 {
     int idx = 0;
     u64 start_ts, end_ts, diff;
-    struct start_data *stdp, std;
+    struct start_data *stdp;
 
     // Handle the case CSenter, Ienter, CSexit, Iexit
     // Handle the case Ienter, CSenter, CSexit, Iexit
@@ -205,22 +210,25 @@ TRACEPOINT_PROBE(preemptirq, TYPE_enable)
     }
 
     stdp = sts.lookup(&idx);
-    bpf_probe_read(&std, sizeof(struct start_data), stdp);
+    if (!stdp) {
+        reset_state();
+        return 0;
+    }
 
     // Handle the case Ienter, Csenter, Iexit, Csexit
-    if (!std.active) {
+    if (!stdp->active) {
         reset_state();
         return 0;
     }
 
     // Handle the case CSenter, Ienter, Iexit, CSexit
-    if (std.idle_skip) {
+    if (stdp->idle_skip) {
         reset_state();
         return 0;
     }
 
     end_ts = bpf_ktime_get_ns();
-    start_ts = std.ts;
+    start_ts = stdp->ts;
 
     if (start_ts > end_ts) {
         reset_state();
@@ -238,8 +246,8 @@ TRACEPOINT_PROBE(preemptirq, TYPE_enable)
     struct data_t data = {};
 
     if (bpf_get_current_comm(&data.comm, sizeof(data.comm)) == 0) {
-        data.addrs[START_CALLER_OFF] = std.addr_offs[START_CALLER_OFF];
-        data.addrs[START_PARENT_OFF] = std.addr_offs[START_PARENT_OFF];
+        data.addrs[START_CALLER_OFF] = stdp->addr_offs[START_CALLER_OFF];
+        data.addrs[START_PARENT_OFF] = stdp->addr_offs[START_PARENT_OFF];
         data.addrs[END_CALLER_OFF] = args->caller_offs;
         data.addrs[END_PARENT_OFF] = args->parent_offs;
```

Reply:
:+1: